qiime2/sdk/usage.py

```diff
@@ -0,0 +1,368 @@
+        for key, val in kwargs.items():
+            if not isinstance(val, str):
+                raise TypeError(
+                    'Name provided for key %r must be a string, not a %r.' %
+                    (key, type(val)))
+
+        self.values = kwargs
+
+    def __repr__(self):
+        return 'UsageOutputNames(**%r)' % (self.values, )
+
+    def get(self, key):
+        return self.values[key]
+
+    def validate(self, signature):
+        provided = set(self.values.keys())
+        exp_outputs = set(signature.outputs)
+
+        missing = exp_outputs - provided
+        if len(missing) > 0:
+            raise ValueError('Missing output(s): %r' % (missing, ))
+
+        extra = provided - exp_outputs
+        if len(extra) > 0:
+            raise ValueError('Extra output(s): %r' % (extra, ))
+
+    def validate_derived(self, derived_outputs):
+        provided = set(derived_outputs.keys())
+        exp_outputs = set(self.values.values())
+
+        missing = exp_outputs - provided
+        if len(missing) > 0:
+            raise ValueError('SDK implementation is missing output(s): %r' %
+                             (missing, ))
+
+        extra = provided - exp_outputs
+        if len(extra) > 0:
+            raise ValueError('SDK implementation has specified extra '
+                             'output(s): %r' % (extra, ))
+
+    def build_opts(self, action_signature, scope):
+        opts = {}
+
+        for output in action_signature.outputs.keys():
+            opts[self.get(output)] = output
+
+        return opts
+
+
+class ScopeRecord:
+    STATE_PENDING = 'pending'
+    STATE_FINISHED = 'finished'
+
+    def __init__(self, id: str, factory: callable = None,
+                 value: object = None,
+                 assert_has_line_matching: callable = None):
+        if factory is not None and not callable(factory):
+            # TODO
+            raise Exception
+
+        if assert_has_line_matching is not None and \
+                not callable(assert_has_line_matching):
+            # TODO
+            raise Exception
+
+        self.id = id
+        if value is None:
+            self.state = self.STATE_PENDING
+        else:
+            self.state = self.STATE_FINISHED
+        self._factory = factory
+        self._result = value
+        self._assert_has_line_matching_ = assert_has_line_matching
+
+    def __repr__(self):
+        if self.state == self.STATE_PENDING:
+            return 'ScopeRecord<id=%s, state=%s>' % (self.id, self.state)
+        else:
+            return 'ScopeRecord<id=%s, result=%r>' % (self.id, self.result)
+
+    @property
+    def result(self):
+        if self.is_pending:
+            raise Exception
+        return self._result
+
+    @property
+    def is_pending(self):
+        return self.state == self.STATE_PENDING
+
+    def resolve(self):
+        if self.is_pending:
+            self._result = self._factory()
+            self.state = self.STATE_FINISHED
```

Comment:
It seems like there's a bit of overlap in function between `_result`/`state` and `result`/`resolve()`. It seems like state could be a computed property of result (if you had a UNSET singleton), and then we could worry less about synchronizing `_result` and `is_pending` for `.result`, also, maybe `.result` and `resolve()` could just be merged, they are both R words which could almost do the same thing.  This might simplify the constructor as well.

Reply:
I need to think a bit about this.